src/js/tracing/reactnativenavigation.ts

```diff
@@ -0,0 +1,206 @@
+   */
+  public registerRoutingInstrumentation(
+    listener: TransactionCreator,
+    beforeNavigate: BeforeNavigate
+  ): void {
+    super.registerRoutingInstrumentation(listener, beforeNavigate);
+
+    this._navigation
+      .events()
+      .registerCommandListener(this._onCommand.bind(this));
+
+    this._navigation
+      .events()
+      .registerComponentWillAppearListener(
+        this._onComponentWillAppear.bind(this)
+      );
+  }
+
+  /**
+   * To be called when a navigation command is dispatched
+   */
+  private _onCommand(): void {
+    if (this._latestTransaction) {
+      this._discardLatestTransaction();
+    }
+
+    this._latestTransaction = this.onRouteWillChange(
+      getBlankTransactionContext(ReactNativeNavigationInstrumentation.name)
+    );
+
+    this._stateChangeTimeout = setTimeout(
+      this._discardLatestTransaction.bind(this),
+      this._options.routeChangeTimeoutMs
+    );
+  }
+
+  /**
+   * To be called AFTER the state has been changed to populate the transaction with the current route.
+   */
+  private _onComponentWillAppear(event: ComponentWillAppearEvent): void {
+    // If the route is a different key, this is so we ignore actions that pertain to the same screen.
+    if (
+      this._latestTransaction &&
+      (!this._prevComponentEvent ||
+        event.componentId != this._prevComponentEvent.componentId)
+    ) {
+      this._clearStateChangeTimeout();
+
+      const originalContext = this._latestTransaction.toContext();
+      const routeHasBeenSeen = this._recentComponentIds.includes(
+        event.componentId
+      );
+
+      const updatedContext = {
+        ...originalContext,
+        name: event.componentName,
+        tags: {
+          ...originalContext.tags,
+          "routing.route.name": event.componentName,
+        },
+        data: {
+          ...originalContext.data,
+          route: {
+            ...event,
+            hasBeenSeen: routeHasBeenSeen,
+          },
+          previousRoute: this._prevComponentEvent,
+        },
+      };
+
+      let finalContext = this._beforeNavigate?.(updatedContext);
+
+      // This block is to catch users not returning a transaction context
+      if (!finalContext) {
+        logger.error(
+          `[${ReactNativeNavigationInstrumentation.name}] beforeNavigate returned ${finalContext}, return context.sampled = false to not send transaction.`
+        );
+
+        finalContext = {
+          ...updatedContext,
+          sampled: false,
+        };
+      }
+
+      if (finalContext.sampled === false) {
+        logger.log(
+          `[${ReactNativeNavigationInstrumentation.name}] Will not send transaction "${finalContext.name}" due to beforeNavigate.`
+        );
+      }
+
+      this._latestTransaction.updateWithContext(finalContext);
+
+      this._prevComponentEvent = event;
+    } else {
+      this._discardLatestTransaction();
+    }
+  }
+
+  /** Cancels the latest transaction so it does not get sent to Sentry. */
```

Comment:
Why did we decide this is the behaviour vs. just finishing the transaction with a different status? Won’t the transaction being cancelled still be valuable to know?

Reply:
Got it, makes sense to me. Should we create a span measuring the time between `_onCommand` and `_onComponentWillAppear`?

Also, can we add your comment this as a comment in the `ReactNativeNavigationInstrumentation` docstring?

Reply:
also, do you think it makes sense to emit breadcrumbs for each command?

Reply:
I agree with both, lets add the both the span and the breadcrumbs in a separate PR so we can add that to all the other routing instrumentations as well.